client/modules/cm-tarnation/src/parser/parser.ts

```diff
@@ -0,0 +1,257 @@
+            for (let i = 0; i < t[4].length; i++) {
+              // const [id, inclusive] = close[i]
+              const c = t[4][i]
+              // const idx = ctx.stack.last(id)
+              const idx = ctx.stack.last(c[0])
+
+              if (idx !== null) {
+                // cut off anything past the closing element
+                // i.e. inside nodes won't persist outside their parent if they
+                // never closed before their parent did
+                ctx.stack.close(idx)
+
+                // if inclusive of the closing token we need to push the token early
+                // if (type && inclusive && !pushed) {
+                if (t[0] && c[1] && !pushed) {
+                  // ctx.buffer.add([type, from, to, 4])
+                  ctx.buffer.add([t[0], t[1], t[2], 4])
+                  ctx.stack.increment()
+                  pushed = true
+                }
+
+                // finally pop the node
+                // const [node, pos, children] = ctx.stack.pop()!
+                const s = ctx.stack.pop()!
+                // ctx.buffer.add([node, pos, inclusive ? to : from, children * 4 + 4])
+                ctx.buffer.add([s[0], s[1], c[1] ? t[2] : t[1], s[2] * 4 + 4])
+                ctx.stack.increment()
+              }
+            }
+          }
+
+          // push the actual token to the buffer, if it hasn't been already
+          // if (type && !pushed) {
+          if (t[0] && !pushed) {
+            // // ctx.buffer.add([type, from, to, 4])
+            ctx.buffer.add([t[0], t[1], t[2], 4])
+            ctx.stack.increment()
+          }
+
+          break
+        }
+      }
+    }
+
+    this.context.index += 1
+  }
+
+  /**
+   * Takes the current parser context, clones it, and then makes sure that
+   * every element on the stack has been resolved. Returns the cloned
+   * context. If the stack is already complete, the current context will be
+   * returned instead.
+   */
+  private finishIncompleteStack() {
+    if (!FINISH_INCOMPLETE_STACKS) return this.context
+    if (!this.context.stack.length) return this.context
+    // temporary clone, we don't want to affect the existing state
+    const ctx = this.context.clone()
+    while (ctx.stack.length) {
+      const [startID, startPos, children] = ctx.stack.pop()!
+      ctx.buffer.add([startID, startPos, this.pos, children * 4 + 4])
+      ctx.stack.increment()
+    }
+    return ctx
+  }
+
+  /**
+   * Fully advances the parser, using the tokenizer's emitted chunks
+   * directly with no additional parsing. This is for debug purposes, as it
+   * allows distinguishing between bugs that orginate from either the
+   * tokenizer or the parser.
+   */
+  advanceFullyRaw(): { buffer: number[]; reused: Tree[] } {
+    const buffer: number[] = []
+    for (let idx = 0; idx < this.pending.length; idx++) {
+      const tokens = this.pending[idx].compile()
+      if (tokens.length === 0) console.error("what")
```

Comment:
Can you add a more useful error message here?

Reply:
whoops missed this console log. removed it in a commit